### PR TITLE
Refactor matching algorithm and tests

### LINKED ISSUE #12: test failure for min_maximal_matching_typical test
======================================================================
FAIL: test_min_maximal_matching_typical (dwave_networkx.algorithms.tests.test_matching.TestMatching)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dwavesystems/dwave_networkx/dwave_networkx/algorithms/tests/test_matching.py", line 435, in test_min_maximal_matching_typical
    "nodes: {}\nedges:{}".format(G.nodes(), G.edges()))
AssertionError: False is not true : nodes: [0, 1, 2, 3, 4, 5, 6]
edges:[(1, 6), (2, 5), (3, 4)]
----------------------------------------------------------------------
Ran 76 tests in 22.659s
FAILED (failures=1)

### LINKED ISSUE #27: Algos/Matching: maximal_matching() not exposed as IF
I suggest adding a comment to maximal_matching() that it's recommended to use a poly-time algo instead so users understand why it's not brought up to the external interface

### PR DESCRIPTION
**Pull request by arcondello**
Refactor minimum maximal matching algorithm and associated tests.

Closes #12
Closes #27

### CODE REVIEW

File: dwave_networkx/algorithms/matching.py [hunk 1]
```diff
@@ -13,10 +13,164 @@
...
+                if u == v:
+                    bqm.linear[u] += 1
+                else:
+                    bqm.add_interaction(u, v, 1)
+
+    return bqm
+
+
+def min_maximal_matching_bqm(G, maximal_lagrange=2, matching_lagrange=None):
+    """Return a BQM with ground states corresponding to a minimum maximal
```
Comment (JoelPasvolsky) [inline dwave_networkx/algorithms/matching.py:137]: Same as above. 

"""Find a binary quadratic model for the graph's minimum maximal matchings.

A matching is a subset of edges in which no node occurs more than
once. A maximal matching is one in which no edges from graph G can be
added without violating the matching rule. This function returns a
binary quadratic model (BQM) with ground states corresponding to the
possible maximal matchings of G.

File: dwave_networkx/algorithms/matching.py [hunk 2]
```diff
@@ -13,10 +13,164 @@
...
+        The graph on which to find a minimum maximal matching.
+
+    maximal_lagrange : float (optional, default=2)
+        The Lagrange multiplier for the maximal constraint. Should be greater
+        than 1.
+
+    matching_lagrange : float (optional)
+        The Lagrange multiplier for the matching constraint. Should be postivie
+        and greater than `maximal_lagrange * max_degree - 2`.
+        Defaults to `1.25 * (maximal_lagrange * max_degree - 2)`.
```
Comment (JoelPasvolsky) [inline dwave_networkx/algorithms/matching.py:152]: Here we're referencing a default set in a different function. If the default changes there, there's a high risk this remains stale. Better to reference the function.

File: dwave_networkx/algorithms/matching.py [hunk 3]
```diff
@@ -13,10 +13,164 @@
...
+
+def maximal_matching_bqm(G, lagrange=None):
+    """Return a BQM with ground states corresponding to the possible maximal
+    matchings of G.
+
+    A matching is a subset of edges in which no node occurs more than
+    once. A maximal matching is one in which no edges from G can be
+    added without violating the matching rule.
+
+    Finding maximal matchings can be done is polynomial time, so this BQM
```
Comment (JoelPasvolsky) [inline dwave_networkx/algorithms/matching.py:84]: "done is polynomial" --> in

File: dwave_networkx/algorithms/matching.py [hunk 4]
```diff
@@ -30,6 +184,9 @@ def maximal_matching(G, sampler=None, **sampler_args):
     once. A maximal matching is one in which no edges from G can be
     added without violating the matching rule.
 
+    Finding maximal matchings can be done is polynomial time, so this method
```
Comment (JoelPasvolsky) [inline dwave_networkx/algorithms/matching.py:187]: "done is polynomial" --> in

File: dwave_networkx/algorithms/matching.py [hunk 5]
```diff
@@ -62,44 +219,22 @@ def maximal_matching(G, sampler=None, **sampler_args):
     References
     ----------
 
-    `Matching on Wikipedia <https://en.wikipedia.org/wiki/Matching_(graph_theory)>`_
+    `Matching on Wikipedia <https://w.wiki/r9s>`_
```
Comment (JoelPasvolsky) [inline dwave_networkx/algorithms/matching.py:222]: Why use the short form here we space is not an issue and the longer form is more readable?

File: dwave_networkx/algorithms/matching.py [hunk 6]
```diff
@@ -13,10 +13,164 @@
...
+def matching_bqm(G):
+    """Return a BQM with ground states corresponding to the possible matchings
+    of G.
+
+    A matching is a subset of edges in which no node occurs more than
+    once.
+
+    Finding valid matchings can be done is polynomial time, so this BQM
+    is only efficient when combined with other constraints and objectives.
+
```
Comment (JoelPasvolsky) [inline dwave_networkx/algorithms/matching.py:42]: efficient or useful?

polynomial time. This BQM
may be useful when combined with other constraints and objectives
to solve a larger hard problem.

File: dwave_networkx/algorithms/matching.py [hunk 7]
```diff
@@ -13,10 +13,164 @@
...
+
+def matching_bqm(G):
+    """Return a BQM with ground states corresponding to the possible matchings
+    of G.
+
+    A matching is a subset of edges in which no node occurs more than
+    once.
+
+    Finding valid matchings can be done is polynomial time, so this BQM
```
Comment (JoelPasvolsky) [inline dwave_networkx/algorithms/matching.py:40]: "done is polynomial" --> in

File: dwave_networkx/algorithms/matching.py [hunk 8]
```diff
@@ -13,10 +13,164 @@
...
+    ----------
+    G : NetworkX graph
+        The graph on which to find a minimum maximal matching.
+
+    maximal_lagrange : float (optional, default=2)
+        The Lagrange multiplier for the maximal constraint. Should be greater
+        than 1.
+
+    matching_lagrange : float (optional)
+        The Lagrange multiplier for the matching constraint. Should be postivie
```
Comment (JoelPasvolsky) [inline dwave_networkx/algorithms/matching.py:150]: "postivie"

File: dwave_networkx/algorithms/matching.py [hunk 9]
```diff
@@ -13,10 +13,164 @@
...
+    Finding maximal matchings can be done is polynomial time, so this BQM
+    is only efficient when combined with other constraints and objectives.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        The graph on which to find a maximal matching.
+
+    lagrange : float (optional)
+        The Lagrange multiplier for the matching constraint. Should be postivie
```
Comment (JoelPasvolsky) [inline dwave_networkx/algorithms/matching.py:93]: "postivie"

**Review by JoelPasvolsky — APPROVED** [review]

File: dwave_networkx/algorithms/matching.py (hunk 6 above)
Comment (arcondello) [reply dwave_networkx/algorithms/matching.py:42]: I think we need to connect those two concepts a bit more explicitly, how about 
```    
    Finding valid matchings can be done in polynomial time, so finding matching
    with BQMs is generally inefficient.
    This BQM may be useful when combined with other constraints and objectives.
```

File: dwave_networkx/algorithms/matching.py (hunk 5 above)
Comment (arcondello) [reply dwave_networkx/algorithms/matching.py:222]: To keep it under 80 characters

File: dwave_networkx/algorithms/matching.py [hunk 10]
```diff
@@ -13,10 +13,164 @@
...
+        and greater than `max_degree - 2`.
+        Defaults to `1.25 * (max_degree - 2)`.
```
[the remark this answers is not in the document]
Comment (arcondello) [reply dwave_networkx/algorithms/matching.py:95]: I think it's readable as-is without needing the additional boilerplate

**Pull request merged**